scale_build/packages/package.py

```diff
@@ -0,0 +1,189 @@
+from scale_build.utils.logger import get_logger
+from scale_build.utils.run import run
+from scale_build.utils.paths import GIT_LOG_PATH, HASH_DIR, PKG_LOG_DIR, SOURCES_DIR
+
+from .binary_package import BinaryPackage
+from .bootstrap import BootstrapMixin
+from .build import BuildPackageMixin
+from .clean import BuildCleanMixin
+from .overlay import OverlayMixin
+from .utils import DEPENDS_SCRIPT_PATH, get_install_deps, normalize_build_depends, normalize_bin_packages_depends
+
+
+logger = logging.getLogger(__name__)
+
+
+class Package(BootstrapMixin, BuildPackageMixin, BuildCleanMixin, OverlayMixin):
+    def __init__(
+        self, name, branch, repo, prebuildcmd=None, kernel_module=False, explicit_deps=None,
+        generate_version=True, predepscmd=None, deps_path=None, subdir=None, deoptions=None, jobs=None,
+        buildcmd=None, tmpfs=True, tmpfs_size=12, batch_priority=100
+    ):
+        self.name = name
+        self.branch = branch
+        self.origin = repo
+        self.prebuildcmd = prebuildcmd or []
+        self.buildcmd = buildcmd or []
+        self.kernel_module = kernel_module
+        self.explicit_deps = set(explicit_deps or set())
+        self.generate_version = generate_version
+        self.predepscmd = predepscmd or []
+        self.deps_path = deps_path
+        self.subdir = subdir
+        self.deoptions = deoptions
+        self.jobs = jobs
+        self.tmpfs = tmpfs
+        self.tmpfs_size = tmpfs_size
+        self.initialized_deps = False
+        self._binary_packages = []
+        self.build_depends = set()
+        self.source_package = None
+        self.parent_changed = False
+        self._build_time_dependencies = None
+        self.build_stage = None
+        self.logger = get_logger(f'{self.name}_package', f'packages/{self.name}.log', 'w')
+        self.children = set()
+        self.batch_priority = batch_priority
+
+    def __eq__(self, other):
+        return other == self.name if isinstance(other, str) else self.name == other.name
+
+    @property
+    def log_file_path(self):
+        return os.path.join(PKG_LOG_DIR, f'{self.name}.log')
+
+    @property
+    def package_path(self):
+        pkg_path = self.source_path
+        if self.subdir:
+            pkg_path = os.path.join(pkg_path, self.subdir)
+        return pkg_path
+
+    @property
+    def debian_control_file_path(self):
+        if self.deps_path:
+            return os.path.join(self.package_path, self.deps_path, 'control')
+        else:
+            return os.path.join(self.package_path, 'debian/control')
+
+    @property
+    def source_path(self):
+        return os.path.join(SOURCES_DIR, self.name)
+
+    @property
+    def binary_packages(self):
+        if self._binary_packages:
+            return self._binary_packages
+
+        if self.name == 'kernel' or (self.predepscmd and not self.deps_path):
+            # We cannot determine dependency of this package because it does not probably have a control file
```

Comment:
What do you mean by grub package not having a control file?

Reply:
If we run `predepscmd` (which I suppose are meant to be run before retrieving dependencies), it will appear